### DISCUSSION TITLE
How to download TestFlight Beta Feedback as CSV Spreadsheet

### DISCUSSION THREAD

**Discussion by PaulSolt** [fastlane/fastlane#18348] [Q&A] [answered]
How do you download TestFlight Beta Feedback as a CSV to open in a spreadsheet?

I ran into problems with the #15793 [TestFlight feedback starter script](https://github.com/fastlane/fastlane/pull/15793#issue-351343917) when I attempted to download feedback.

* It appears that the `feedback.tester` can be `nil` so the script crashes with some feedback responses. It appears that Apple no longer populates these fields with anything other than "Anonymous" for the name. 
* You can get an email using the field: `feedback.email_address`

**Comment by PaulSolt** [comment, marked as answer]
You can export the TestFlight Beta Feedback by using this lane that will create a .csv file. Opening the .csv file in Numbers makes it super easy to tap on an email to respond to a tester or clicking on a screenshot to get insight into the problem area.
 
Edit your Fastfile to include the following lane.

**TestFlight Beta Feedback**

```
lane :connect_feedback do
  fastlane_require 'spaceship'
  fastlane_require 'csv'
  
  identifier = CredentialsManager::AppfileConfig.try_fetch_value(:app_identifier)
  team_id = CredentialsManager::AppfileConfig.try_fetch_value(:team_id)
  itc_team_id = CredentialsManager::AppfileConfig.try_fetch_value(:itc_team_id)
  
  Spaceship::Tunes.login
  # Spaceship::Tunes.select_team

  Spaceship::Tunes.client.team_id = itc_team_id

  # Gets app
  app = Spaceship::ConnectAPI::App.find(identifier)

  # Gets feedback for an app (default includes screenshots and tester info)
  feedbacks = app.get_beta_feedback

  headers = ["Date", "Email", "Comment", "Build", "Device", "iOS Version", "Screenshot"]

  CSV.open("reviews.csv", "w") do |csv|
    csv << headers
    feedbacks.each do |feedback|
      # device_model os_version
      response = ["#{feedback.timestamp}", "#{feedback.email_address}", "#{feedback.comment}", "#{feedback.build.version}", "#{feedback.device_model}", "#{feedback.os_version}"]
      
      # Iterate over feedback screenshots
      feedback.screenshots.each do |screenshot|
        image_asset = screenshot.image_assets.first        
        response += ["#{image_asset["url"]}"]
      end
      
      csv << response
    end
  end
end
```

## Download TestFlight Beta Feedback

Run the `connect_feedback` script using:

```bundle exec fastlane connect_feedback```

## AppStore Connect Credentials Required  

You'll need to setup your `fastlane/Appfile` with your credentials, or modify the script above with your team / bundle identifier.

## Easily Respond or Dig into Problems

Now you can easily open the output reviews.csv in Numbers, click emails to send an email, or click on URLs to see the screenshot.

<img width="1181" alt="Brew Coffee beta feedback" src="https://user-images.githubusercontent.com/371902/131859526-e61cc08e-a19e-493e-a671-403f9e8decf3.png">

**Reply by kcw-grunt** [reply]
I have tried to use: `feedbacks = app.get_beta_feedback` and searched the Spaceship repo and there is no method at all...there is no reference (circa Sep 2022). It fails.

Is there a version change I'm unaware of @PaulSolt ?
|Screenshot of Error|
|-------------------|
|![Screen Shot 2022-09-18 at 9 19 11 AM](https://user-images.githubusercontent.com/2899463/190892744-4bcc1f93-0558-4f7d-8eb5-c4a19ead9e04.png)|

**Reply by lukas-ruzicka** [reply]
@kcw-grunt The script was removed from the fastlane. I've done some work on a tool that fetches and manages the feedback (now it is creating GitHub Issues out of it) - you can check it out here: https://github.com/lukas-ruzicka/testflight-feedback . Hope it helps!

**Reply by PaulSolt** [reply]
Nice work @lukas-ruzicka 

@kcw-grunt I wasn't aware this was removed, but I haven't used it for a while.

**Comment by mylogon341** [comment]
I had some issues with this to begin with, but I changed the `feedbacks` initialiser to 
`feedbacks = Spaceship::ConnectAPI.get_beta_feedback(filter: {'build.app' => <app id here>}, includes: "tester,build,screenshots", sort: "-timestamp")`
The only thing I've not been able to figure out yet is how to avoid having to type in the username, so if you have any suggestions, that'd be grand.

**Reply by mylogon341** [reply]
I have now solved this and slimmed this script down a bit
```ruby
lane :connect_feedback do
    fastlane_require 'spaceship'
    fastlane_require 'csv'
    
    Spaceship::ConnectAPI.login(user = "<email address>")
    # Not a problem as I'm only on one team so this will auto select
    Spaceship::ConnectAPI.select_team
  
    # Gets feedback for an app (default includes screenshots and tester info)
    feedbacks = Spaceship::ConnectAPI.get_beta_feedback(filter: {'build.app' => "<app id>"}, includes: "tester,build,screenshots", sort: "-timestamp")
  
    headers = ["Date", "Email", "Comment", "Build", "Device", "iOS Version", "Screenshot"]
  
    CSV.open("reviews.csv", "w") do |csv|
      csv << headers
      feedbacks.each do |feedback|
        # device_model os_version
        response = ["#{feedback.timestamp}", "#{feedback.email_address}", "#{feedback.comment}", "#{feedback.build.version}", "#{feedback.device_model}", "#{feedback.os_version}"]
        
        # Iterate over feedback screenshots
        feedback.screenshots.each do |screenshot|
          image_asset = screenshot.image_assets.first        
          response += ["#{image_asset["url"]}"]
        end
        
        csv << response
      end
    end
  end
```

I hope this helps someone down the line

**Reply by kselvocki** [reply]
@mylogon341 - Thank you _so much_ for this! I'd been trying to figure out a workaround over the past couple weeks, and then I took another look at this thread yesterday and saw your update.

You may or may not know the answer to this, but any idea how to get it to pull more than 50 responses? I don't see any params for that, but I'm also extremely new to fastlane and ruby.

Again, seriously, thank you. Spotting your update here legit made my week.

**Reply by mylogon341** [reply]
Oh cool. Really glad that I could help – that's made my week :)

As for the 50 limit, that's something that I've not come across yet as we're only up to 47 so I guess I would have come across this next week for sure.

Looking at the docs quickly [here](https://www.rubydoc.info/gems/fastlane/2.161.0/Spaceship%2FConnectAPI%2FApp:get_beta_feedback), it looks as it takes an optional input parameter for `limit`, which probably internally defaults to 50. I'd have a go at putting that in and setting it to a much larger number and seeing If you can get away with it. Something like this

```ruby
feedbacks = Spaceship::ConnectAPI.get_beta_feedback(filter: {'build.app' => "<app id>"}, includes: "tester,build,screenshots", limit: 1000, sort: "-timestamp")
```

I'll also try that too next week so my spreadsheet doesn't magically stop growing.

Have a great weekend 👋

**Reply by kselvocki** [reply]
Yes, that's the one! Tested it out and the max for the `limit` parameter is 2000, but that's definitely an improvement over 50. I'll dive into the docs and play around a little more next week: currently, even with the limit set to 2000, my export is only collecting 200. Either way, I'm in a _much_ better position than I was yesterday (and my coworker who was having to manually download all of the TF feedback is extremely grateful). Thank you again--have a great weekend!

**Comment by ChrisFeldmeier** [comment]
Hi guys, I created https://testflight-control.onlyvision.app/, maybe someone will find it useful :)
